### PR TITLE
Support threads

### PR DESCRIPTION
**Pull request by richfromm**
This adds support for threads.

Previously, all messages, even if they originated in a thread, were posted directly to the main channel.
Now we only post the non-threaded messages to the main channel.
If there is a thread, create a thread at that message, then post the messages in the thread.
Calls to `sorted()` ensure that both the messages in the main channel, and those within a thread, are posted in order.
I suspect this is not needed, as the JSON file (in limited tests) appears to be ordered, but it doesn't hurt.

Rather than repeat the JSON file parsing, we instead parse it just once, when the file is input, and build up a dict with the relevant contents. See the docstring of `parse_json_slack_export()` for details. The actual bot implementation no longer references the file, and instead iterates through this dict.

There are still issues with the blocking nature of the GUI and it hanging after entering the token, but resolving that is out of scope.
To compensate, there are bare print statements added in the bot command method (`slackord(ctx)`) for now, since nothing output to the GUI in that method is ever seen.

### CODE REVIEW

**Comment by richfromm** [comment]
This PR may not be the best place to discuss it, but regarding the issues with the GUI hanging, how set are you on even having a GUI? Is your intention that you only need to enter the token once, then the advantage of the GUI is that you can repeatedly choose per-channel JSON files, then use the bot to import the data to the appropriate channels?

I'm not convinced that there's that much of an advantage of having a GUI at all. There's only two items to specify: the input JSON file, and the token. That can easily be done on the command line. The token wouldn't even need to be required, good alternatives are in a dot file (which could be in the cwd, or the user's home dir), or via an environment variable.

I'm also not convinced that using a bot at all is the best way to do this.

You can set up a client with:
```
intents = discord.Intents.default()
intents.messages = True
intents.message_content = True
client = discord.Client(intents=intents)
```

Set up the asyncio properly (I'm still trying to figure out the best way to do that), then send a specific message with:
```
    channel = client.get_channel(channel_id)
    await channel.send('this is a test')
```

This seems a bit simpler to me than having a bot involved. The only downside I can think of is that there's no easy way to query the channel id given the channel name. The best method I've figured out so far is right clicking on the channel name in the Discord GUI, and that has a prereq of enabling Developer Mode.

Thoughts? I'd like to fix the GUI hanging issue, but the easiest way to do so might be to get rid of the GUI. I can understand if that's not your preference, or if you don't want to deal with any significant changes to this and would rather support https://github.com/thomasloupe/Slackord2 instead, but if that's the case then I might consider forking, perhaps renaming "slackordcli" and pointing to you as the original upstream source.

**Comment by thomasloupe** [comment]
There's really no necessity for a GUI, however, just keep in mind that this entire project is just something I created in my free time and shared for others' benefit. I'm happy to merge this into master. I think the changes here benefit everyone.

There is one change that should be made, which is once the threads are created, and the last message in the thread is sent, the thread should be immediately archived so additional responses cannot be recorded inside the thread. They close after 24 hours, and while it seems like a neat idea to be able to interact with past messages, Slackord is meant to export historical data, not allow interactions with historical data.

If that's something you can tidy up when there is time, I can go ahead and merge this. The review looks fine.

**Comment by thomasloupe** [comment]
As far as GUI hanging is concerned, I had a look into this a couple years back, the issue with thread hanging is definitely from the bot taking over the main thread in which Tkinter needs to continue drawing the GUI. Threading tasks in Python was not something I wanted to venture into as it was quite a bit more complicated than doing it in C#/.NET.

Other than looks, the ease of use of copy/pasting a token, and providing basic "About" information for the app, there is no real reason for a GUI other than wanting to learn how to write Python Tkinter. Removing the GUI would be fine as long as the same information from the GUI is available in the CLI.

**Comment by richfromm** [comment]
Apologies if this all ought to be obvious, but I'm new to Discord.

Are you sure that threads actually close after 24 hours, as opposed to just being hidden?

While I do indeed see that described in the docs at https://support.discord.com/hc/en-us/articles/4403205878423-Threads-FAQ :

> You may choose the duration of your thread with the Archive After Inactivity drop-down menu:

I haven't yet been able to find that in the Discord GUI. The closest I can find is:

> Hide After Inactivity
> 24 Hours
> New threads will not show in the channel list after being inactive for the specified duration.

Where can I find this? And is there a global setting I can change to disable it?

I suppose I can see an argument in favor of keeping conversation flowing in the main channel, but it seems a little odd to me. I have definitely experienced many situations in the past in Slack of significantly old threads getting new input.

In my particular use case, I'm migrating an entire Slack to Discord, and the Discord is starting out empty. I'm not sure if I see the advantage of preventing new conversation on old threads. The old vs. new will still be very clear, as the old messages are posted as the bot, rather than as the actual users, and the user info is just contained within the text. Same with the timestamps. I'd rather if you could actually fake the timestamp and user to be what they were in slack, but I'm pretty sure that's impossible.

What do you think of making it an option to keep the thread open or immediately close it? (which could default to close)

File: README.md [hunk 1]
```diff
@@ -10,38 +10,43 @@ Alternatively, you can download the slackord.py script here and run it directly
...
+    ```
+    brew install python-tk.
+    ```
+
+    Windows:
+    ```
+    curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py
+    python get-pip.py
+    pip install discord.py
+    ```
```
Comment (richfromm) [inline README.md:47]: the indentation allows us to keep the second half of the bulletted list below auto numbered and have it do the right thing, as opposed to start over in counting a new list

**Comment by thomasloupe** [comment]
See: [How Do I Archive a Thread](https://support.discord.com/hc/en-us/articles/4403205878423-Threads-FAQ#h_01F99ZAK4XF1JQZFAQFKNXT46W)

When a thread is archived by an admin or moderator, that thread cannot be unarchived unless another admin or moderator unarchives it. When threads are archived by lifecycle expirations, anyone can reply and reopen the thread. There is a ridiculous amount of work involved with giving the option for both, which I believe is out of scope for Slackord, probably even Slackord2, but given the explanation, I don't see a problem with leaving them open by default, since typically the idea is that folks who are in the channel reading the threads already have access to read/reply to the threads in that channel if it were Slack. Continuing that user experience and expected functionality on Discord could be less jarring.

**Comment by richfromm** [comment]
So this video (from 2021) shows how to get to this setting:
https://www.youtube.com/watch?v=916rPqvwpvM

But when I go to the server settings on my discord server, there's nothing for "Channels"
This is via the web, not the app. I wonder if that matters.

Anyway, this was my my first experimentaton with threads, I think I'll just wait a little bit and see if the threads I've created get closed. (Is it obvious in the GUI when a thread is closed?)

Regardless, it seems like it's easy to unarchive a thread:
https://support.discord.com/hc/en-us/articles/4403205878423-Threads-FAQ#h_01F9A1VV0J73QKZQXJ6YZ24YBD

So I don't have a problem closing (I'm assuming "close" is the same as "archive", and "open" is the same as "unarchive"?) the thread after the import.

**UPDATE:** Oh, just saw your message above, and the diff as to whether or not the thread is archived by an admin or moderator, vs. lifecycle expirations. I'm fine archiving as long as you don't need to be a moderator to unarchive. I wonder what the behavior is if using the API? Let me look into that...

**Comment by richfromm** [comment]
I'm not actually sure if it's possible to archive a thread via the API.

You can query whether or not it's archived:
https://discordpy.readthedocs.io/en/latest/api.html?highlight=send#discord.Thread.archived

But I don't see methods for either archiving or unarchiving:
https://discordpy.readthedocs.io/en/latest/api.html?highlight=send#thread

Unless maybe you just manually edit `thread.archived` ?

I'll explore that.

**Comment by thomasloupe** [comment]
If using the API, you would probably need to pass some type of user/scope, so it'd either be defined as Bot, Admin, or Moderator (typically). I am actually not entirely sure it's possible or even if discord.py would have that functionality. 

Instead of exploring that possibility, I think it is fine to let the Admin of the Discord specify the timeout manually (defaulted to 24hr). This is, by far, enough control. If Admins or Moderators wish to archive threads, they can do so manually if they wish.

To change the "Hide After Inactivity" for threads:
Right-click a text channel>Edit Channel>Overview>Scroll to the very bottom>Hide After Inactivity.
Example:
image

**Comment by richfromm** [comment]
yes, I found "Hide After Inactivity"
is that the same thing as archiving?
that's the only use I've found of the word "Hide". I was expecting "Archive After Inactivity"

**Comment by thomasloupe** [comment]
Yep, that's it.

**Comment by thomasloupe** [comment]
I believe the verbiage for "Hide" is used because archived threads get put on their own tab after being archived, which can then later be searched. I don't necessarily agree with the wording, but it is accurate to what they actually do.
See example:
image

**Comment by richfromm** [comment]
I'm able to query `created_thread.archived` (it's `False`), but my test as to whether setting it to `True` has the desired effect is running into an issue in that nothing is working for me at all any more, I keep getting a 401: `discord.errors.LoginFailure: Improper token has been passed.`

I even tried resetting the token, and am still getting it.

Am I maybe getting rate limited, and it's returning that rather than a 429?

Maybe this is a sign that I should take a break and get back to this sometime later.

My tentative preference is that if setting `created_thread.archived = True` works, and has a result that anyone can unarchive the thread by posting to it again, that seems like a reasonable thing to do to me.

And if not, then don't worry about it.

I'll let you know if/when I can get things working again.

**Comment by thomasloupe** [comment]
Are you using the Bot token or an OAuth token? API calls from outside of the bot won't use the bot token, but instead pass the scopes and auth from the OAuth side of things.

**Comment by richfromm** [comment]
Okay, tried one last time immediately after posting that above, and now I'm able to run fine again, no 401. Hmmm.

Anyway, I'm setting the boolean:
```
Message posted: 1661639877.272379
Message posted: 1661639887.608799
Message in thread posted: 1661639899.256209
Message in thread posted: 1661639902.518509
Message in thread posted: 1661639924.453549
BEFORE: archived for the thread is False
AFTER: archived for the thread is True
Message posted: 1661639892.134999
Message posted: 1661639913.404409
Message posted: 1661639928.049759
Done posting messages
```

But I see no indication in the GUI of the threads actually being archived.

Googling a bit, the Javascript Discord docs say:

> To archive or unarchive a thread, use the `ThreadChannel#setArchived()` method and pass in a boolean parameter:

https://discordjs.guide/popular-topics/threads.html#archiving-unarchiving-and-locking-threads
https://discord.js.org/#/docs/main/stable/class/ThreadChannel?scrollTo=setArchived

But I don't see any such `ThreadChannel` in the Python docs:

https://discordpy.readthedocs.io/en/latest/search.html?q=threadchannel

I'm inclined to say that as long as the threads posted by the bot get auto archived in the usual manner (I can wait a day and verify that that happens), then this isn't worth worrying about any further.

**Comment by richfromm** [comment]
To answer your question, I'm pretty sure it's an OAuth token.

Anyway, it's working again.

I'm going to take a break now, not sure exactly when I'll be back online. I'll see if the test threads I all posted today get archived tomorrow.

**Comment by richfromm** [comment]
I did some tests, and **no** threads that I created and posted to in discord ever got automatically closed. This is both just using the GUI by hand, as well as with the bot. With the default 24 hour timestamp, as well as with changing that to only 1 hour.

I'm not sure why, but I also don't think I want to spend too much energy exploring this tangent. I do see that if I manually close a thread, that **is** visible in the UI, so it's not just a question of me not being able to see what's going on. (Unless closing and archiving are different things? But I thought the summary from above is that they are synonymous.)

So I was originally going to suggest that I'd add a comment (and/or something in the README) that any threads that get posted via the bot will be closed in the usual manner based on age, but since I can't actually demonstrate that, I think maybe it's better to not say anything at all.

The tl;dr is that I think this is fine as is, and am requesting that you merge it, unless you have any specific additional requests.

**Comment by thomasloupe** [comment]
Hey @richfromm, I am good with merging this.
I wanted to point out that there's a possibility that you haven't seen the thread closed because you have the access level to reply to the thread to keep it open. Perhaps if you view the thread from a different user's role, it would be closed and archived?

However, this is just an afterthought, and doesn't really mean much in the grand scheme of things regarding the ability to allow threaded replies.

I'm going ahead with the merge.

**Comment by thomasloupe** [comment]
Sidenote, even though this is closed: I've added your contribution to the Slackord python script itself in the comments at the top, and have released the newest build with the bumped version.

Appreciate your contribution to Slackord and making it better, cheers!
